rel/value_set_array.go

```diff
@@ -1,40 +1,321 @@
+func (a Array) Kind() int {
+	return arrayKind
+}
+
+// Bool returns true iff the tuple has attributes.
+func (a Array) IsTrue() bool {
+	if len(a.values) == 0 {
+		panic("Empty string not allowed (should be == None)")
+	}
+	return true
+}
+
+// Less returns true iff v is not a number or tuple, or v is a tuple and t
+// precedes v in a lexicographical comparison of their name/value pairs.
+func (a Array) Less(v Value) bool {
+	if a.Kind() != v.Kind() {
+		return a.Kind() < v.Kind()
+	}
+	b := v.(Array)
+	if a.offset != b.offset {
+		return a.offset < b.offset
+	}
+	n := len(a.values)
+	if n < len(b.values) {
+		n = len(b.values)
+	}
+	for i, av := range a.values[:n] {
+		bv := b.values[i]
+		if av.Less(bv) {
+			return true
+		}
+		if bv.Less(av) {
+			return false
+		}
+	}
+	return len(a.values) < len(b.values)
+}
+
+// Negate returns {@neg: b}.
+func (a Array) Negate() Value {
+	return NewTuple(NewAttr(negateTag, a))
+}
+
+// Export exports an Array as a slice.
+func (a Array) Export() interface{} {
+	result := make([]interface{}, 0, a.Count())
+	for _, v := range a.values {
+		result = append(result, v.Export())
+	}
+	return result
+}
+
+// Count returns the number of elements in the Array.
+func (a Array) Count() int {
+	return len(a.values)
+}
+
+// Has returns true iff the given Value is in the Array.
+func (a Array) Has(value Value) bool {
+	if t, ok := value.(ArrayItemTuple); ok {
+		if a.offset <= t.at && t.at < a.offset+len(a.values) {
+			return t.item == a.values[t.at-a.offset]
+		}
+	}
+	return false
+}
+
+func (a Array) with(index int, item Value) Set {
+	if a.index(index) == len(a.values) {
+		return Array{values: append(a.values, item), offset: a.offset}
+	} else if index == a.offset-1 {
+		return Array{
+			values: append(append(make([]Value, 0, 1+len(a.values)), item), a.values...),
+			offset: a.offset - 1,
+		}
+	}
+	return newSetFromSet(a).With(NewArrayItemTuple(index, item))
+}
+
+// With returns the original Array with given value added. Iff the value was
+// already present, the original Array is returned.
+func (a Array) With(value Value) Set {
+	if t, ok := value.(ArrayItemTuple); ok {
+		return a.with(t.at, t.item)
+	}
+	return newSetFromSet(a).With(value)
+}
+
+// Without returns the original Array without the given value. Iff the value
+// was already absent, the original Array is returned.
+func (a Array) Without(value Value) Set {
+	if t, ok := value.(ArrayItemTuple); ok {
+		if i := a.index(t.at); i >= 0 && t.item == a.values[i] {
+			if t.at == a.offset {
+				return Array{values: a.values[1:], offset: a.offset + 1}
+			}
+			if t.at == a.offset+i {
+				return Array{values: a.values[:i], offset: a.offset}
+			}
+			return newSetFromSet(a).Without(value)
+		}
+	}
+	return a
+}
+
+// Map maps values per f.
+func (a Array) Map(f func(v Value) Value) Set {
+	result := NewSet()
+	for e := a.Enumerator(); e.MoveNext(); {
+		result = result.With(f(e.Current()))
+	}
+	return result
```

Comment:
nit: `result` should be a `[]Value` and the function returns `NewSet(result...)`

Reply:
Map is defined over the elements of the set (that is, `(@: ..., @item: ...)`), not the items of the array.

Reply:
I meant something like this
```go
        result := make([]Value)
	for e := a.Enumerator(); e.MoveNext(); {
		result = append(result, f(e.Current()))
	}
	return NewSet(result...)
```

but after looking at `NewSet` implementation, apparently it's the same thing.